### PR TITLE
OOP-ify Riem_Solver_C

### PR DESCRIPTION
**Pull request by eddie-c-davis**
## Purpose

This PR OOP-ifies the `Riem_Solver_C` code.

TODO: Some variables need doc string info.

## Code changes:

- Add `RiemannSolverC` class.
- Modify `DynCore` to call `_riem_solver_c` object.

## Checklist
Before submitting this PR, please make sure:

- [x] You have followed the coding standards guidelines established at [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--BD7zigBMAhMZAPkeNENeuU2UAg-IlsYffZgTwyKEylty7NhY).
- [x] Docstrings and type hints are added to new and updated routines, as appropriate

### CODE REVIEW

**Comment by eddie-c-davis** [comment]
launch jenkins

**Review by oelbert — APPROVED** [review]
Looks good! Thanks for handling this one!

File: fv3core/stencils/riem_solver_c.py [hunk 1]
```diff
@@ -57,89 +53,113 @@ def finalize(
...
+        ptc: FloatField,
+        q_con: FloatField,
+        delpc: FloatField,
+        gz: FloatField,
+        pef: FloatField,
+        ws: FloatFieldIJ,
+    ):
+        """
+        Solves for the nonhydrostatic terms for vertical velocity (w)
+        and non-hydrostatic pressure perturbation after D-grid winds advect
```
Comment (rheacangeo) [inline fv3core/stencils/riem_solver_c.py:122]: This one is done after the C-grid winds are advected

File: fv3core/stencils/riem_solver_c.py [hunk 2]
```diff
@@ -57,89 +53,113 @@ def finalize(
...
+           cappa: ??? (in)
+           ptop: pressure at top of atmosphere (in)
+           hs: ??? (in)
+           w3: vertical velocity of the lowest level (in)
+           ptc: potential temperature (in)
+           q_con: total condensate mixing ratio (in)
+           delpc: vertical delta in pressure (in)
+           gz: geopotential heigh (inout)
+           pef: full hydrostatic pressure(inout)
+           ws: vertical velocity (inout)
```
Comment (rheacangeo) [inline fv3core/stencils/riem_solver_c.py:136]: I think the descriptions of these are reversed, the IJ one should be the vertical velocity of the lowest layer. In the riem_solver3 PR I switched the order they are called so the inout would be last, but the order in the __call__ method here is the other way -- sorry for introducing that confusions

File: fv3core/stencils/riem_solver_c.py [hunk 3]
```diff
@@ -41,13 +38,12 @@ def precompute(
     with computation(PARALLEL), interval(0, -1):
         dz = gz[0, 0, 1] - gz
     with computation(PARALLEL), interval(...):
-        gm = 1.0 / (1.0 - cp3)
-        dm = dm / constants.GRAV
+        gm = 1.0 / (1.0 - cappa)
+        dm /= constants.GRAV
```
[the remark this answers is not in the document]
Comment (eddie-c-davis) [reply fv3core/stencils/riem_solver_c.py:42]: For sure!

File: fv3core/stencils/riem_solver_c.py [hunk 4]
```diff
@@ -57,89 +53,113 @@ def finalize(
...
-
     with computation(PARALLEL):
         with interval(0, 1):
             pef = ptop
         with interval(1, None):
             pef = pe2 + pem
     with computation(BACKWARD):
         with interval(-1, None):
-            gz = hs_0
+            gz = hs
```
[the remark this answers is not in the document]
Comment (eddie-c-davis) [reply fv3core/stencils/riem_solver_c.py:63]: Yes, am grateful to @rheacangeo for the `TODO` comments regarding 2D storages.

File: fv3core/stencils/riem_solver_c.py (hunk 1 above)
Comment (eddie-c-davis) [reply fv3core/stencils/riem_solver_c.py:122]: D'oh! Good catch, I meant to change that.

File: fv3core/stencils/riem_solver_c.py (hunk 2 above)
Comment (eddie-c-davis) [reply fv3core/stencils/riem_solver_c.py:136]: Not at all, I will match yours.

**Pull request merged**